go/pkg/client/tokens.go

```diff
@@ -0,0 +1,186 @@
+func requestToken(handshakeClient flight.FlightService_HandshakeClient, handshakeReq *flight.HandshakeRequest) ([]byte, error) {
+	err := handshakeClient.Send(handshakeReq)
+
+	if err != nil {
+		return nil, err
+	}
+
+	handshakeResp, err := handshakeClient.Recv()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return handshakeResp.Payload, nil
+}
+
+// tokenResp protects the current session token (or an error in getting the session token).
+type tokenResp struct {
+	Lock  sync.Mutex
+	Token []byte
+	Error error
+}
+
+// getToken returns the current token, or an error if an error has occurred at some point.
+func (tk *tokenResp) getToken() ([]byte, error) {
+	tk.Lock.Lock()
+	defer tk.Lock.Unlock()
+	if tk.Error != nil {
+		return nil, tk.Error
+	} else {
+		token := make([]byte, len(tk.Token))
+		copy(token, tk.Token)
+		return token, nil
+	}
+}
+
+// setToken sets the session token to a new value.
+func (tk *tokenResp) setToken(tok []byte) {
+	tk.Lock.Lock()
+	tk.Token = tok
+	tk.Error = nil
+	tk.Lock.Unlock()
+}
+
+// setError sets an error value for the session token.
+func (tk *tokenResp) setError(err error) {
+	tk.Lock.Lock()
+	tk.Error = err
+	tk.Lock.Unlock()
+}
+
+// A tokenManager stores the current client token and sends periodic keepalive messages to refresh the client token.
+type tokenManager struct {
+	token *tokenResp // the actual client token, which gets periodically updated.
+	close func() error
+}
+
+// getToken returns the current token, or an error if an error has occurred at some point.
+func (tr *tokenManager) getToken() ([]byte, error) {
+	return tr.token.getToken()
+}
+
+// withToken attaches the current session token to a context as metadata.
+func (tr *tokenManager) withToken(ctx context.Context) (context.Context, error) {
+	tok, err := tr.getToken()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return withAuthToken(ctx, tok), nil
+}
+
+func (tr *tokenManager) Close() error {
+	return tr.close()
+}
+
+// newTokenManager creates a tokenManager that begins a background goroutine that continually refreshes
+// the token so that it does not time out.
+//
+// authString is the authorization string used to get the first token.  Examples:
+//   - "Anonymous" is used for anonymous authentication.
+//   - "io.deephaven.authentication.psk.PskAuthenticationHandler <password>" is used for PSK authentication
+func newTokenManager(ctx context.Context, fs *flightStub, cfg configpb2.ConfigServiceClient, authString string) (*tokenManager, error) {
+	handshakeClient, err := fs.handshake(withAuth(ctx, authString))
+
+	if err != nil {
+		return nil, err
+	}
+
+	tkn, err := requestToken(handshakeClient, &flight.HandshakeRequest{Payload: []byte(authString)})
+
+	if err != nil {
+		return nil, err
+	}
+
+	ac, err := cfg.GetConfigurationConstants(withAuthToken(ctx, tkn), &configpb2.ConfigurationConstantsRequest{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	sessionDurationStr, ok := ac.ConfigValues["http.session.durationMs"]
```

Comment:
Suggest moving magic string literals like this (and "Anonymous", elsewhere) to some utility class so they can be defined in one place